File: dotnet-desktop-guide/winforms/advanced/snippets/how-to-retrieve-data-from-the-clipboard/net/vb/MainForm.vb
Imports System.Collections.Specialized
Imports System.Drawing
Imports System.Windows.Forms

Namespace RetrieveClipboardData

    Public Class MainForm
        Inherits Form

        '<helper-methods>
        <Serializable()>
        Public Class Customer

            Private nameValue As String = String.Empty

            Public Sub New(ByVal name As String)
                nameValue = name
            End Sub

            Public Property Name() As String
                Get
                    Return nameValue
                End Get
                Set(ByVal value As String)
                    nameValue = value
                End Set
            End Property

        End Class
        '</helper-methods>

        Public Sub New()
            Clipboard.Clear()
        End Sub

        '<retrieve-custom-format>
        ' Demonstrates TryGetData using a custom format name and a business object.
        ' Note: In .NET 10, SetData for objects is no longer supported,
        ' so this example shows how to retrieve data that might have been
        ' set by other applications or earlier .NET versions.
        Public ReadOnly Property TestCustomFormat() As Customer
            Get
                Dim data As Object = Nothing
                ' For demonstration, we'll use string data instead of objects
                ' since SetData for objects is no longer supported in .NET 10
                If Clipboard.TryGetData("CustomerFormat", data) Then
                    Return TryCast(data, Customer)
                End If
                Return Nothing
            End Get
        End Property
        '</retrieve-custom-format>

        '<retrieve-multiple-formats>
        ' Demonstrates how to retrieve data from the Clipboard in multiple formats
        ' using TryGetData instead of the obsoleted GetData method.
        Public Sub TestClipboardMultipleFormats()

            Dim dataObject As IDataObject = Clipboard.GetDataObject()

            If dataObject IsNot Nothing Then

                ' Check for custom format
                If dataObject.GetDataPresent("CustomFormat") Then

                    Dim customData As Object = Nothing
                    If Clipboard.TryGetData("CustomFormat", customData) Then

                        Dim item As ListViewItem = TryCast(customData, ListViewItem)
                        If item IsNot Nothing Then
                            MessageBox.Show(item.Text)
                        ElseIf TypeOf customData Is String Then
                            MessageBox.Show(CStr(customData))
                        End If

                    End If

                End If

                ' Check for Customer type - note that object serialization
                ' through SetData is no longer supported in .NET 10
                If dataObject.GetDataPresent(GetType(Customer)) Then

                    Dim customerData As Object = Nothing
                    If Clipboard.TryGetData(GetType(Customer).FullName, customerData) Then

                        Dim customer As Customer = TryCast(customerData, Customer)
                        If customer IsNot Nothing Then
                            MessageBox.Show(customer.Name)
                        End If

                    End If

                End If

                ' For modern .NET 10 applications, prefer using standard formats
                If Clipboard.ContainsText() Then
                    Dim text As String = Clipboard.GetText()
                    MessageBox.Show($"Text data: {text}")
                End If

            End If

        End Sub
        '</retrieve-multiple-formats>

        '<retrieve-common-format>
        ' Demonstrates TryGetData methods for common formats.
        ' These methods are preferred over the older Get* methods.
        Public Function SwapClipboardAudio(ByVal replacementAudioStream As System.IO.Stream) As System.IO.Stream

            Dim returnAudioStream As System.IO.Stream = Nothing

            If Clipboard.ContainsAudio() Then
                returnAudioStream = Clipboard.GetAudioStream()
                Clipboard.SetAudio(replacementAudioStream)
            End If

            Return returnAudioStream

        End Function

        ' Demonstrates TryGetData for file drop lists
        Public Function SwapClipboardFileDropList(ByVal replacementList As StringCollection) As StringCollection

            Dim returnList As StringCollection = Nothing

            If Clipboard.ContainsFileDropList() Then
                returnList = Clipboard.GetFileDropList()
                Clipboard.SetFileDropList(replacementList)
            End If

            Return returnList

        End Function

        ' Demonstrates TryGetData for images
        Public Function SwapClipboardImage(ByVal replacementImage As Image) As Image

            Dim returnImage As Image = Nothing

            If Clipboard.ContainsImage() Then
                returnImage = Clipboard.GetImage()
                Clipboard.SetImage(replacementImage)
            End If

            Return returnImage

        End Function

        ' Demonstrates TryGetData for text in HTML format
        Public Function SwapClipboardHtmlText(ByVal replacementHtmlText As String) As String

            Dim returnHtmlText As String = Nothing

            If Clipboard.ContainsText(TextDataFormat.Html) Then
                returnHtmlText = Clipboard.GetText(TextDataFormat.Html)
                Clipboard.SetText(replacementHtmlText, TextDataFormat.Html)
            End If

            Return returnHtmlText

        End Function

        ' Example of using TryGetData for custom string-based data
        Public Function GetCustomStringData(ByVal format As String) As String

            Dim data As Object = Nothing
            If Clipboard.TryGetData(format, data) Then
                Return TryCast(data, String)
            End If

            Return Nothing

        End Function
        '</retrieve-common-format>

    End Class

End Namespace

File: dotnet-desktop-guide/winforms/advanced/snippets/how-to-retrieve-data-from-the-clipboard/net/csharp/MainForm.cs
using System.Collections.Specialized;
using System.Drawing;
using System.Windows.Forms;

namespace RetrieveClipboardData;

public class MainForm : Form
{
    //<HelperMethods>
    [Serializable]
    public class Customer
    {
        private string nameValue = string.Empty;
        public Customer(string name)
        {
            nameValue = name;
        }
        public string Name
        {
            get { return nameValue; }
            set { nameValue = value; }
        }
    }
    //</HelperMethods>

    public MainForm()
    {
        Clipboard.Clear();
    }

    //<RetrieveCustomFormat>
    // Demonstrates TryGetData using a custom format name and a business object.
    // Note: In .NET 10, SetData for objects is no longer supported,
    // so this example shows how to retrieve data that might have been
    // set by other applications or earlier .NET versions.
    public Customer? TestCustomFormat
    {
        get
        {
            // For demonstration, we'll use string data instead of objects
            // since SetData for objects is no longer supported in .NET 10
            if (Clipboard.TryGetData("CustomerFormat", out object? data))
            {
                return data as Customer;
            }
            return null;
        }
    }
    //</RetrieveCustomFormat>

    //<RetrieveMultipleFormats>
    // Demonstrates how to retrieve data from the Clipboard in multiple formats
    // using TryGetData instead of the obsoleted GetData method.
    public void TestClipboardMultipleFormats()
    {
        IDataObject? dataObject = Clipboard.GetDataObject();
        
        if (dataObject != null)
        {
            // Check for custom format
            if (dataObject.GetDataPresent("CustomFormat"))
            {
                if (Clipboard.TryGetData("CustomFormat", out object? customData))
                {
                    if (customData is ListViewItem item)
                    {
                        MessageBox.Show(item.Text);
                    }
                    else if (customData is string stringData)
                    {
                        MessageBox.Show(stringData);
                    }
                }
            }

            // Check for Customer type - note that object serialization
            // through SetData is no longer supported in .NET 10
            if (dataObject.GetDataPresent(typeof(Customer)))
            {
                if (Clipboard.TryGetData(typeof(Customer).FullName!, out object? customerData))
                {
                    if (customerData is Customer customer)
                    {
                        MessageBox.Show(customer.Name);
                    }
                }
            }

            // For modern .NET 10 applications, prefer using standard formats
            if (Clipboard.ContainsText())
            {
                string text = Clipboard.GetText();
                MessageBox.Show($"Text data: {text}");
            }
        }
    }
    //</RetrieveMultipleFormats>

    //<RetrieveCommonFormat>
    // Demonstrates TryGetData methods for common formats.
    // These methods are preferred over the older Get* methods.
    public Stream? SwapClipboardAudio(Stream replacementAudioStream)
    {
        Stream? returnAudioStream = null;
        if (Clipboard.ContainsAudio())
        {
            returnAudioStream = Clipboard.GetAudioStream();
            Clipboard.SetAudio(replacementAudioStream);
        }
        return returnAudioStream;
    }

    // Demonstrates TryGetData for file drop lists
    public StringCollection? SwapClipboardFileDropList(StringCollection replacementList)
    {
        StringCollection? returnList = null;
        if (Clipboard.ContainsFileDropList())
        {
            returnList = Clipboard.GetFileDropList();
            Clipboard.SetFileDropList(replacementList);
        }
        return returnList;
    }

    // Demonstrates TryGetData for images
    public Image? SwapClipboardImage(Image replacementImage)
    {
        Image? returnImage = null;
        if (Clipboard.ContainsImage())
        {
            returnImage = Clipboard.GetImage();
            Clipboard.SetImage(replacementImage);
        }
        return returnImage;
    }

    // Demonstrates TryGetData for text in HTML format
    public string? SwapClipboardHtmlText(string replacementHtmlText)
    {
        string? returnHtmlText = null;
        if (Clipboard.ContainsText(TextDataFormat.Html))
        {
            returnHtmlText = Clipboard.GetText(TextDataFormat.Html);
            Clipboard.SetText(replacementHtmlText, TextDataFormat.Html);
        }
        return returnHtmlText;
    }

    // Example of using TryGetData for custom string-based data
    public string? GetCustomStringData(string format)
    {
        if (Clipboard.TryGetData(format, out object? data))
        {
            return data as string;
        }
        return null;
    }
    //</RetrieveCommonFormat>
}